core/src/main/java/com/linecorp/armeria/server/file/StreamingHttpFile.java

```diff
@@ -0,0 +1,222 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.file;
+
+import java.io.Closeable;
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpResponseWriter;
+import com.linecorp.armeria.common.MediaType;
+
+public abstract class StreamingHttpFile<T extends Closeable> extends AbstractHttpFile {
+
+    private static final Logger logger = LoggerFactory.getLogger(StreamingHttpFile.class);
+    private static final int MAX_CHUNK_SIZE = 8192;
+
+    protected StreamingHttpFile(@Nullable MediaType contentType,
+                                boolean dateEnabled,
+                                boolean lastModifiedEnabled,
+                                @Nullable BiFunction<String, HttpFileAttributes, String> entityTagFunction,
+                                HttpHeaders headers) {
+        super(contentType, dateEnabled, lastModifiedEnabled, entityTagFunction, headers);
+    }
+
+    @Override
+    protected final HttpResponse doRead(HttpHeaders headers, long length,
+                                        Executor fileReadExecutor) throws IOException {
+        final T in = newStream();
+        if (in == null) {
+            return null;
+        }
+
+        final HttpResponseWriter res = HttpResponse.streaming();
+        res.write(headers);
+        fileReadExecutor.execute(() -> doRead(res, in, 0, length, fileReadExecutor));
+        return res;
+    }
+
+    private void doRead(HttpResponseWriter res, T in, long offset, long end, Executor fileReadExecutor) {
+        final int chunkSize = (int) Math.min(MAX_CHUNK_SIZE, end - offset);
+        final byte[] buf = new byte[chunkSize];
+        final int readBytes;
+        try {
+            readBytes = read(in, buf);
+        } catch (Exception e) {
+            close(res, in, e);
+            return;
+        }
+
+        if (readBytes < 0) {
+            // Should not reach here because we only read up to the end of the stream.
+            // If reached, it may mean the stream has been truncated.
+            close(res, in, new EOFException());
+            return;
+        }
+
+        if (readBytes > 0) {
+            if (!res.tryWrite(HttpData.of(buf, 0, readBytes))) {
```

Comment:
I'm curious if we could `ByteBufHttpData` here, because this object can be released by Armeria without a user's care. But there might be a problem which requires another `read` method ...

Reply:
Yeah, I added `HttpFile.aggregateWithPooledObject()` and `HttpFile.read()` now accepts `ByteBufAllocator` to create `ByteBufHttpData` instances.